Preparations to make Java Session and Flash immutable

mkurz opened the pull request:
Java's `Session` and `Flash` are mutable right now... extending `HashMap`...

Let's make it like the Scala ones.
That's what they will look like after Play 2.7 when all the deprecations are removed: https://gist.github.com/mkurz/6878b1f8b0b7a352fcd4683385faff06
(`getOptional` will be renamed back to `get`)

mkurz commented on framework/src/play/src/main/java/play/mvc/Http.java:
```diff
@@ -2133,24 +2136,74 @@ public void discardCookie(String name, String path, String domain, boolean secur
+         * @deprecated Deprecated as of 2.7.0.
+         */
+        @Deprecated
         public boolean isDirty = false;
```
The "dirty" flags are only used in `JavaHelpers.createResult` which is deprecate in Play 2.7 and will be removed anyway: https://github.com/playframework/playframework/blob/88abf01f673c7400e933f663ab0903cf5617454c/framework/src/play/src/main/scala/play/core/j/JavaHelpers.scala#L147-L160

mkurz commented on framework/src/play/src/main/java/play/mvc/Security.java:
```diff
@@ -93,10 +90,10 @@ public String getUsername(Context ctx) {
          */
-        public String getUsername(Request req) {
-            return req.session().get("username");
+        public Optional<String> getUsername(Request req) {
```
This method was just added during 2.7 development therefore we can just change the return type now.

marcospereira commented:
I think this makes sense, although I'm not so sure about overriding and deprecating `Map` methods. One thing that is mildly annoying is that we don't have good/consistent names for APIs that manipulates headers, session, flash, and cookies. For example, we have `discardCookies`, `removingFromSession` and `clearlingLang`. And in some other places just `remove` or `-`.

Here is a (not extensive) compilation:

```
play.api.mvc.Session {
  def get(key: String): Option[String]
  def +(kv: (String, String)): Session
  def -(key: String): Session
  def apply(key: String): String = data(key)
}

play.api.mvc.Flash {
  def get(key: String): Option[String]
  def +(kv: (String, String)): Flash
  def -(key: String): Flash
  def apply(key: String): String
}

play.api.mvc.Result {
    def withHeaders(headers: (String, String)*)
    def withDateHeaders(headers: (String, ZonedDateTime)*)
    def discardingHeader(name: String)
    def withCookies(cookies: Cookie*)
    def discardingCookies(cookies: DiscardingCookie*)
    def withSession(session: Session)
    def withSession(session: (String, String)*)
    def withNewSession
    def flashing(flash: Flash)
    def flashing(values: (String, String)*)
    def as(contentType: String)
    def session(implicit request: RequestHeader): Session = newSession getOrElse request.session
    def addingToSession(values: (String, String)*)(implicit request: RequestHeader)
    def removingFromSession(keys: String*)(implicit request: RequestHeader)
}

play.mvc.Result {
    public Optional<String> header(String header)
    public Map<String, String> headers()
    public Result withFlash(Flash flash)
    public Result withFlash(Map<String, String> flash)
    public Result withNewFlash()
    public Result flashing(Map<String, String> values)
    public Result flashing(String key, String value)
    public Result removingFromFlash(String... keys)
    public Session session()
    public Session session(Http.Request request)
    public Result withSession(Session session)
    public Result withSession(Map<String, String> session)
    public Result withNewSession()
    public Result addingToSession(Http.Request request, Map<String, String> values)
    public Result addingToSession(Http.Request request, String key, String value)
    public Result removingFromSession(Http.Request request, String... keys)
    public Cookie cookie(String name)
    public Optional<Cookie> getCookie(String name)
    public Cookies cookies()
    public Result withCookies(Cookie... newCookies) {
    public Result discardCookie(String name)
    public Result discardCookie(String name, String path)
    public Result discardCookie(String name, String path, String domain)
    public Result discardCookie(String name, String path, String domain, boolean secure)
    public Result withHeader(String name, String value)
    public Result withHeaders(String... nameValues)
    public Result discardHeader(String name)
    public Result withLang(Lang lang, MessagesApi messagesApi)
    public Result clearingLang(MessagesApi messagesApi)
}

play.api.mvc.Request {
    def queryString: Map[String, Seq[String]] = target.queryMap
    def headers: Headers
    def withHeaders(newHeaders: Headers): RequestHeader =
    def attrs: TypedMap
    def withAttrs(newAttrs: TypedMap): RequestHeader =
    def addAttr[A](key: TypedKey[A], value: A): RequestHeader =
    def removeAttr(key: TypedKey[_]): RequestHeader =
    def getQueryString(key: String): Option[String] = target.getQueryParameter(key)
    def cookies: Cookies = attrs(RequestAttrKey.Cookies).value
    def session: Session = attrs(RequestAttrKey.Session).value
    def flash: Flash = attrs(RequestAttrKey.Flash).value
    def rawQueryString: String = target.queryString
    def withTransientLang(lang: Lang): RequestHeader =
    def withTransientLang(code: String): RequestHeader =
    def withTransientLang(locale: Locale): RequestHeader =
    def clearTransientLang(): RequestHeader =
    def transientLang(): Option[Lang] =
}

play.mvc.Request {
    TypedMap attrs();
    RequestHeader withAttrs(TypedMap newAttrs);
    <A> RequestHeader addAttr(TypedKey<A> key, A value);
    RequestHeader removeAttr(TypedKey<?> key);
    Map<String,String[]> queryString();
    String getQueryString(String key);
    Cookies cookies();
    Cookie cookie(String name);
    default Session session()
    default Flash flash()
    Headers getHeaders();
    default Optional<String> header(String headerName)
    default boolean hasHeader(String headerName) {
    default RequestHeader withTransientLang(Lang lang) {
    default RequestHeader withTransientLang(String code) {
    default RequestHeader withTransientLang(Locale locale) {
    default RequestHeader clearTransientLang() {
    default Optional<Lang> transientLang() {
}

play.api.mvc.Headers {
  def hasHeader(headerName: String): Boolean = get(headerName).isDefined
  def add(headers: (String, String)*): Headers = new Headers(this.headers ++ headers)
  def apply(key: String): String = get(key).getOrElse(scala.sys.error("Header doesn't exist"))
  def get(key: String): Option[String] = getAll(key).headOption
  def getAll(key: String): Seq[String] = toMap.getOrElse(key, Nil)
  def keys: Set[String] = toMap.keySet
  def remove(keys: String*): Headers = {
  def replace(headers: (String, String)*): Headers = remove(headers.map(_._1): _*).add(headers: _*)
}

play.mvc.Http.Headers {
    public Map<String, List<String>> toMap()
    public boolean contains(String headerName)
    public Optional<String> get(String name)
    public List<String> getAll(String name)
    public Headers addHeader(String name, String value)
    public Headers addHeader(String name, List<String> values)
    public Headers remove(String name)
}
```

Okay.. I'm going too much out of scope here. Anyway, I think it may also be a good chance to have consistent names.

marcospereira commented on framework/src/play/src/main/java/play/mvc/Http.java:
```diff
@@ -2184,6 +2258,185 @@ public void clear() {
+        // ### Let's deprecate all of HashMap
+
+        /**
+         * @deprecated Deprecated as of 2.7.0.
```
Better to give some explanation too.

marcospereira commented on framework/src/play/src/main/java/play/mvc/Http.java:
```diff
@@ -2193,24 +2446,74 @@ public void clear() {
             super(data);
         }
 
+        public Flash(play.api.mvc.Flash underlying) {
```
Comments made about Session also applies here.

marcospereira commented on framework/src/play/src/main/java/play/mvc/Result.java:
```diff
@@ -269,9 +269,7 @@ public Result flashing(Map<String, String> values) {
-            Map<String, String> newValues = new HashMap<>(this.flash);
-            newValues.putAll(values);
-            return withFlash(newValues);
+            return withFlash(this.flash.adding(values));
```
👍

mkurz commented:
@marcospereira I addresses all your comments, please have a look again. I used `remove` instead of `removing` now. We probably want to align all the name variations one day... In a different pull request :wink:

mkurz commented:
:+1: Thanks, I was almost sure you don't like `$plus$plus` and `$plus` :laughing:
